src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.PartialReload.cs

```diff
@@ -0,0 +1,169 @@
+﻿#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Uno.Extensions;
+using Uno.Foundation.Logging;
+using Uno.UI.RemoteControl.HotReload.Messages;
+using System.Runtime.Loader;
+using System.Runtime.CompilerServices;
+using System.Diagnostics;
+
+namespace Uno.UI.RemoteControl.HotReload
+{
+	partial class ClientHotReloadProcessor
+	{
+		private Dictionary<string, Type>? _mappedTypes;
+
+		private bool _supportsLightweightHotReload;
+
+		private Task? _updatingTypes;
+		private readonly object _updatingTypesGate = new();
+
+		private void InitializePartialReload()
+		{
+			_supportsLightweightHotReload = (_msbuildProperties?.TryGetValue("TargetFramework", out var targetFramework) ?? false)
+				&& (_msbuildProperties?.TryGetValue("BuildingInsideVisualStudio", out var buildingInsideVisualStudio) ?? false)
+				&& buildingInsideVisualStudio.Equals("true", StringComparison.OrdinalIgnoreCase)
+				&& (
+					targetFramework.Contains("-android")
+					|| targetFramework.Contains("-ios"));
+
+			_mappedTypes = _supportsLightweightHotReload
+				? BuildMappedTypes()
+				: new();
+		}
+
+		private async Task PartialReload(FileReload fileReload)
+		{
+			if (!_supportsLightweightHotReload)
+			{
+				if (this.Log().IsEnabled(LogLevel.Debug))
+				{
+					this.Log().Debug($"Skipping file reload");
+				}
+
+				return;
+			}
+
+			if (!fileReload.IsValid())
+			{
+				if (fileReload.FilePath is not null && this.Log().IsEnabled(LogLevel.Debug))
+				{
+					this.Log().LogDebug($"FileReload is missing a file path");
+				}
+
+				if (fileReload.Content is null && this.Log().IsEnabled(LogLevel.Debug))
+				{
+					this.Log().LogDebug($"FileReload is missing content");
+				}
+
+				return;
+			}
+
+			lock (_updatingTypesGate)
+			{
+				if (_updatingTypes == null || _updatingTypes.Status is TaskStatus.RanToCompletion or TaskStatus.Faulted)
+				{
+					_updatingTypes = ObserveUpdateTypeMapping();
+				}
+				else
+				{
+					if (this.Log().IsEnabled(LogLevel.Debug))
+					{
+						this.Log().Debug($"LightweightReload: Waiting for existing type observer");
+					}
+				}
+			}
+
+			await _updatingTypes;
+		}
+
+		private async Task ObserveUpdateTypeMapping()
+		{
```

Comment:
Do we need to prevent re-entrance here for the scenario where multiple files are modified in within the 15 second monitoring window?

Reply:
This is handled by the `await _updatingTypes;` above. There can be only one awaiter for this.